vulnerabilities/models.py

```diff
@@ -110,6 +111,38 @@ class VulnerabilityReference(models.Model):
     def scores(self):
         return VulnerabilitySeverity.objects.filter(reference=self.id)
 
+    def save(self, *args, **kwargs):
+        if self.id:
+            super(VulnerabilityReference, self).save(*args, **kwargs)
+        else:
+            url_parsed = urlpy.parse(self.url)
+            self.url = str(url_parsed.canonical())
+            url_scheme = url_parsed.scheme
+            scheme_independent_url = self.url[len(url_scheme) :]
+            if url_scheme == "http":
+                similar_instance = VulnerabilityReference.objects.filter(
+                    vulnerability=self.vulnerability,
+                    source=self.source,
+                    reference_id=self.reference_id,
+                    url="https" + scheme_independent_url,
+                ).first()
+                if not similar_instance:
+                    super(VulnerabilityReference, self).save(*args, **kwargs)
+            elif url_scheme == "https":
+                similar_instance = VulnerabilityReference.objects.filter(
+                    vulnerability=self.vulnerability,
+                    source=self.source,
+                    reference_id=self.reference_id,
+                    url="http" + scheme_independent_url,
+                ).first()
+                if similar_instance:
+                    similar_instance.url = self.url
+                    similar_instance.save()
+                else:
+                    super(VulnerabilityReference, self).save(*args, **kwargs)
+            else:
+                super(VulnerabilityReference, self).save(*args, **kwargs)
```

Comment:
Same as above, for the computers it doesn't make difference, but when reading I don't have to verify whether the return value is correct.

Reply:
> @savish28 I'm not really sure how much of impact this is going to have on the db side things, since we are doing more queries per `save`. The imports are already slow enough.
> 
> So here's what I suggest. Keep the url normalization part in the `save` method. The scheme thing, should be moved into a separate job/command, https://docs.djangoproject.com/en/3.1/howto/custom-management-commands/ .

@sbs2001 So, a nice idea, Does this mean that the command/job `python manage.py job_name` have to be called explicitly by the admin to run the following job and it would go through all the entries and it would update the db. I think it can be extended in two ways:-
1. If no arguments are passed then all the entries are looked upon and are checked.
2. If arguments like id/Vulnerability/source etc are passed and only filtered objects are processed.

I think that in this case, each run would cost O(N) queries and would not be automated.

But in the save method update, it would cost 1 extra query per update and would be automated. So I think query wise the override save method is more time-efficient and more automated.

What would be your suggestion in the case and if we are going with the job idea, does the proposed method would work?

Reply:
@savish28  

>called explicitly by the admin

We could get clever by running the job after each import, so this is a non-issue, I guess. 

>1. If no arguments are passed then all the entries are looked upon and are checked.
  2. If arguments like id/Vulnerability/source etc are passed and only filtered objects are processed.
    
Btw I want to remove the source thing when I get time :) it's not populated anywhere and is a bad way to keep track of source(s) .  With that out of the way, what would be the use case of (2) ? In almost all cases (1) would be used. 

> each run would cost O(N) queries

IMHO that can be done `1 + (no of duplicates) `. One query to fetch all references (just the id and url columns) . Iterate over all records from server side, check for dups in the already fetched records. If a dup is found follow the procedure along the lines of what you're doing here, so that's gonna cost 1 `DELETE`  query per dup. 

Dups are rare anyways so it won't be O(N) . 

> it would cost 1 extra query per update

IMHO that'd be 1 extra query per insert/ (`save` method invocation in most cases )

Reply:
Ohh okay, I get the idea. We are importing all the data and then doing this job. This would result in more efficiency rather than doing operation at each save.

So I guess these are the deliverables now,
- [ ] Create Job for removing duplicates.
- [ ] Add functionality to run the job after each import.
Does this seem right? @sbs2001